drivers: stepper: DRV8424 stepper driver

jbehrensnx opened the pull request:
This adds a step/dir stepper driver for the [drv8424](https://www.ti.com/product/DRV8424) stepper motor driver.
It uses a counter to ensure accurate timing of the step signal.

### Tests

The driver comes with two test suites:
1. **stepper_drv8424_api**, which checks both the drivers conformity to the stepper api and expected return values. This test suite supports the Nucleo-767ZI and MIMXRT1060-EVKB boards as well as the native_sim target. The test suite was run on all three of these platforms
2. **stepper_drv8424_emul**, which uses the gpio_emul driver to evaluate the values of the gpio pins used by this driver. It currently only test the enable api command. This test suite only supports the native_sim target, on which it was run.

### Considerations during Development
PWM was discarded during development, as using pwm for driving the step pin prevents accurate counting of steps. This limits pwm to the run mode and it was felt that using two different technologies for driving the step pin added unnecessary complexity and introduced possible incompatibilities with some gpio controllers that did not allow remapping of the pin between gpio and pwm.

The counter api was chosen for generating the step signal as it allows for both a time accurate signal (within the limits of delays introduced by interrupt handling) and for accurate counting of the number of steps taken, necessary for move_by and move_to. This driver requires that the counter supports both setting a new top value as well as registering a callback for counter wraparound.

The consideration for time accuracy is also the reason why the step pin needs to be connected to a gpio controller that is directly on the SOC, as data transfer via bus can lead to jitter and an thus unstable signal motor movement. Another reason for the gpio controller on the SOC is the fact that this driver calls gpio api commands in an interrupt context, which is unsafe over a bus.

### Future Work

- Development of a drv8424 driver that implements gradual acceleration for move_to/move_by and run
- Implementation of an RPI Pico counter driver hat allows the usage of this driver on RP2040 boards

fabiobaltieri commented:
heya, this should probably re-use the common step/dir driver code, see `drivers/stepper/step_dir_stepper_common.c`, also it should definitely not use dynamic memory, could you look into it? `adi_tmc22xx_stepper_controller.c` is probably a good point of reference

Irockasingranite commented:
Hi! Co-author of this driver here.

> this should probably re-use the common step/dir driver code

We specifically chose a hardware counter as the driving source for steps, because we found that the workqueue based implementation that is now `step_dir_stepper_common` leads to noticeable discrepancy between the set and the actual speed at higher frequencies (>1000 steps/s). By using a counter to generate interrupts at very regular intervals, this is mitigated (at the cost of taking up the counter, and doing some work in interrupts).

Ideally we would like to separate that part into an alternative step-dir implementation that is reusable (and selectable) between drivers, but only after getting feedback on the implementation in the context of a specific driver, and an RFC on what such a mechanism to select between different underlying implementations might look like.

> also it should definitely not use dynamic memory

Dynamic allocation is used to offload event callbacks to a workqueue, since they are user registered and may not be safe to call in an interrupt context. If no dynamic allocation is a requirement, this can probably be changed to use something like a message queue instead, at the cost of some extra static allocation.

faxe1008 commented:
Having an alternate counter based implementation one could switch to via Kconfig would be lovely.

fabiobaltieri commented:
> Kconfig has the issue that you can not configure the step-dir implementation on an per-instance basis which is an issue if you use multiple instances of the same driver chip for different tasks.

Fair enough, could you do it devicetree based then? You can have the common config/data with a bunch of const pointer for both options and then only instantiate the one for the configuration you want, and check which ones are not NULL to decide what to do in runtime. There's few drivers using that pattern in input, https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/input/input_gpio_kbd_matrix.c#L302-L329 for example, then you have any combination with multiple instances. Would that work? Having the option on common code would be brilliant.

fabiobaltieri commented:
> Dynamic allocation is used to offload event callbacks to a workqueue, since they are user registered and may not be safe to call in an interrupt context. If no dynamic allocation is a requirement, this can probably be changed to use something like a message queue instead, at the cost of some extra static allocation.

Yeah if you could get away with something statically allocated I think it'd be better, normally we push back dynamic memory on general purpose drivers unless there's a very good reason for it, this can potentially be useful on very small footprint devices, it'd be a waste to require a heap on those.

Irockasingranite commented:
> > Kconfig has the issue that you can not configure the step-dir implementation on an per-instance basis which is an issue if you use multiple instances of the same driver chip for different tasks.
> 
> Fair enough, could you do it devicetree based then? You can have the common config/data with a bunch of const pointer for both options and then only instantiate the one for the configuration you want, and check which ones are not NULL to decide what to do in runtime. There's few drivers using that pattern in input, https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/input/input_gpio_kbd_matrix.c#L302-L329 for example, then you have any combination with multiple instances. Would that work? Having the option on common code would be brilliant.

Yes, I think something like that could work. We could have a dts enum property enumerating all "backends" for the step-dir interface, and then have the common init macros set a list of function pointers depending on its value (like a mini API struct), plus pointers to backend specific config/data. Really this would be sort-of replicating the way `struct device` handles drivers, nested on a smaller scale.

jilaypandya commented on drivers/stepper/CMakeLists.txt:
```diff
@@ -5,6 +5,7 @@ zephyr_syscall_header(${ZEPHYR_BASE}/include/zephyr/drivers/stepper.h)
 
 # zephyr-keep-sorted-start
 add_subdirectory_ifdef(CONFIG_STEPPER_ADI_TMC adi_tmc)
+add_subdirectory_ifdef(CONFIG_STEPPER_ADI_TMC ti)
```
CONFIG_STEPPER_TI :-)

jbehrensnx commented on dts/bindings/stepper/ti/ti,drv8424.yaml:
```diff
@@ -0,0 +1,47 @@
+    type: phandle-array
+    description: Fault pin.
+
+  en-gpios:
```
No, the only gpios defined in stepper-controller.yaml are step-gpios and dir gpios

fabiobaltieri commented:
> Yes, I think something like that could work. We could have a dts enum property enumerating all "backends" for the step-dir interface, and then have the common init macros set a list of function pointers depending on its value (like a mini API struct), plus pointers to backend specific config/data. Really this would be sort-of replicating the way `struct device` handles drivers, nested on a smaller scale.

Yeah that's the idea, there's few examples of "internal apis" in the tree that works like that, the keyboard matrix one, m_can, regulators... I think it's a good pattern.

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+		counter_stop(config->counter);
+		if (data->event_callback != NULL) {
+			/* Ignore return value since we can't do anything about it anyway */
+			drv8424_schedule_user_callback(data, STEPPER_EVENT_STEPS_COMPLETED);
```
I think i understand the motivation behind using `msg_q` here, but i would have expected this not to be part of the driver, as soon as, one of this events is triggered, the driver just `notifies` via the stepper_event_callback_fn and then the queuing mechanism could happen in the application code, or in some sort of `motion_control` subsystem.

Irockasingranite replied:
The concern here is that the registered callback may not be safe to call in an interrupt context. Because we are triggering the `STEPPER_EVENT_STEPS_COMPLETED` event from within the interrupt generated by the counter, we move it into a (workqueue) thread context, where it presumably is safe to call (although it's still possible to construct deadlocks if one tries, since it runs on the system workqueue).

If the API specified that callbacks registered via `stepper_set_event_callback` have to be interrupt safe, this would not be necessary and we could call it directly.

fabiobaltieri commented:
One more thing, the project does not use fixup commits, makes it way hard to review, the orignal stack was fine (one commit for the driver and one for the test), just rebase the fixup back into the original commits and force push

Irockasingranite committed:
drivers: stepper: Add driver for DRV8424 stepper motor controller

Adds a step/dir stepper driver for the drv8424 stepper driver.

Signed-off-by: Sarah Renkhoff <sarah.renkhoff@navimatix.de>

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+	ret = gpio_pin_configure_dt(&config->m0_pin, GPIO_OUTPUT_INACTIVE);
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to reset m0_pin (error: %d)", dev->name, ret);
+		drv8424_microstep_recovery(dev);
```
It's best if this fallback mechanism is exposed to user via include/zephyr/drivers/stepper/stepper_drv8424.h, If the setting of the microstep fails then return -EIO and then the user can decide to use this fallback mechanism

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+{
+	struct drv8424_data *data = dev->data;
+
+	data->actual_position = position;
```
`data->reference_position`

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+			LOG_ERR("%s: Failed to start counter (error %d)", dev->name, ret);
+			data->is_moving = false;
+			goto end;
+		}
```
I believe setting `data->is_moving` at other places can be replaced with 
```suggestion
		} else {
		    data->moving = true;
		}    
```

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+		break;
+	case STEPPER_MICRO_STEP_256:
+		m0_value = 1;
+		m1_value = 2;
```
How about a LUT :-) For indexing you could use `MICRO_STEP_RES_INDEX(res)` from `stepper.h`

jbehrensnx replied:
As C has no easy way to check or enforce that a value of a function parameter that is defined as an enum is actually contained in said enum, a check like this would still be necessary. Otherwise, invalid micro-step resolutions would be interpreted as valid ones (e.g. micro-step resolution 5 would be interpreted as micro-step resolution 4). 
One solution would be to ensure `MICRO_STEP_RES_INDEX` or an equivalent function returns an error code on an invalid micro-step resolution, but that is probably beyond the scope of this PR.

jilaypandya commented on drivers/stepper/ti/Kconfig.drv8424:
```diff
@@ -0,0 +1,21 @@
+	bool "TI DRV8424 stepper motor driver"
+	default y
+	depends on DT_HAS_TI_DRV8424_ENABLED
+	depends on COUNTER
```
select COUNTER

jilaypandya commented on drivers/stepper/ti/Kconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024 Navimatix GmbH
+# SPDX-License-Identifier: Apache-2.0
+
+config STEPPER_TI
```
```suggestion
menuconfig STEPPER_TI
```
this will put all the TI drivers on the next page

jbehrensnx replied:
This causes the compliance check to fail, as STEPPER_TI has no children. The first TI driver that adds corresponding KCONFIG entries should probably change it.

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/CMakeLists.txt:
could you organize the tests like stepper/drv8424/api, the same for emul tests as  well.

jilaypandya replied:
The API tests can be very well used for all the drivers. so how about this,

In overlays, we create something like this,

```
/ {
	zephyr,user {
		motors = <&motor_1>, <&motor_2>;
	};
};
```

And then within tests

```
struct stepper_fixture {
	const struct device *dev[DT_PROP_LEN(DT_PATH(zephyr_user), motors)];
	stepper_event_callback_t callback;
};

static void stepper_before(void *f)
{
	struct stepper_fixture *fixture = f;
	for(uint8_t motor=0;motor<DT_PROP_LEN(DT_PATH(zephyr_user), motors);motor++)
	{
		printk("setting reference position to 0\n for %s",fixture->dev[motor]->name);
		(void)stepper_set_reference_position(fixture->dev[motor], 0);

	}
}


ZTEST_F(stepper, test_target_position)
{
	for (uint8_t motor = 0; motor < DT_PROP_LEN(DT_PATH(zephyr_user), motors); motor++)
	{
		int32_t pos = 100u;
		k_poll_signal_reset(&stepper_signal);

		(void)stepper_set_max_velocity(fixture->dev[motor], 100u);

		/* Pass the function name as user data */
		(void)stepper_set_event_callback(fixture->dev[motor], fixture->callback, &fixture);

		(void)stepper_move_to(fixture->dev[motor], pos);

		(void)k_poll(&stepper_event, 1, K_SECONDS(5));
		unsigned int signaled;
		int result;

		k_poll_signal_check(&stepper_signal, &signaled, &result);
		zassert_equal(signaled, 1, "Signal not set");
		zassert_equal(result, STEPPER_EVENT_STEPS_COMPLETED, "Signal not set");
		(void)stepper_get_actual_position(fixture->dev[motor], &pos);
		zassert_equal(pos, 100u, "Target position should be %d but is %d", 100u, pos);
		zassert_equal(user_data_received, &fixture, "User data not received");
	}
}

```

jbehrensnx replied:
Making the drv8424 tests usable for other drivers is something that I have planned, but is outside the scope of this PR as this will need to include discussions about the intended behavior of api functions. In addition, the stepper drivers would need to be adapted to conform to the behavior expected by these tests afterwards.

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_STEPPER_LOG_LEVEL_DBG=y
+CONFIG_STEPPER=y
+CONFIG_POLL=y
+CONFIG_COUNTER=y
```
This can be dropped after `SELECT_COUNTER` in driver kconfig.

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/src/main.c:
```diff
@@ -0,0 +1,501 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "zephyr/drivers/gpio.h"
```
#include <zephyr/drivers/gpio.h>, same for the other zephyr includes as well :-)

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/boards/mimxrt1060_evkb.overlay:
```diff
@@ -0,0 +1,64 @@
+		 compatible = "ti,drv8424";
+
+
+		 dir-gpios = < &arduino_header 9 0 >; /* D3 */
```
```suggestion
		 dir-gpios = <&arduino_header 9 0>; /* D3 */
```
same for other bindings as well

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+		return ret;
+	}
+
+	switch (m1_value) {
```
Code duplication over here, probably refactor this out in a separate function.

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+	/* Lock interrupts while modifying counter settings */
+	int key = irq_lock();
+
+	/* Set counter to correct frequency */
```
Please elaborate what `correct frequency` means in this context or this comment can also be droppped.

jbehrensnx replied:
The new explanation should make it clearer

jilaypandya replied:
Thanks :-)

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+
+	data->is_moving = true;
+
+	/* Start counter */
```
```suggestion
```

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+	}
+
+	if (!data->enabled) {
+		return -ENODEV;
```
This check should be there, however there has to be another errno. Probably [ECANCELLED](https://docs.zephyrproject.org/apidoc/latest/group__system__errno.html#ga9532d840ef91fd8e1ecc5d7ca7cbf212) would be a more suitable option.

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+		return -ENODEV;
+	}
+
+	/* Set direction pin */
```
```suggestion
```

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+	/* Set device back pointer */
+	data->dev = dev;
+
+	/* Configure direction pin */
```
```suggestion
```

jilaypandya commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,775 @@
+	struct drv8424_data *const data = dev->data;
+	int ret;
+
+	/* Set device back pointer */
```
```suggestion
```

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/src/main.c:
```diff
@@ -0,0 +1,501 @@
+struct k_poll_signal stepper_signal;
+struct k_poll_event stepper_event;
+
+struct gpio_dt_spec en_pin = GPIO_DT_SPEC_GET_OR(DT_NODELABEL(drv8424), en_gpios, {0});
```
```suggestion
```
unused

jilaypandya commented on tests/drivers/stepper/stepper_drv8424_api/src/main.c:
```diff
@@ -0,0 +1,501 @@
+struct k_poll_event stepper_event;
+
+struct gpio_dt_spec en_pin = GPIO_DT_SPEC_GET_OR(DT_NODELABEL(drv8424), en_gpios, {0});
+struct gpio_dt_spec slp_pin = GPIO_DT_SPEC_GET_OR(DT_NODELABEL(drv8424), sleep_gpios, {0});
```
```suggestion
```
unused

bjarki-andreasen commented on dts/bindings/stepper/ti/ti,drv8424.yaml:
```diff
@@ -0,0 +1,62 @@
+    type: phandle-array
+    description: Sleep pin (active low).
+
+  m0-gpios:
+    type: phandle-array
+    description: Microstep configuration pin 0.
+
+  m1-gpios:
+    type: phandle-array
+    description: Microstep configuration pin 1.
```
These are required according to the driver so should be required here as well `required: true`

bjarki-andreasen commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+		.step_pin = GPIO_DT_SPEC_INST_GET(inst, step_gpios),                               \
+		.sleep_pin = GPIO_DT_SPEC_INST_GET_OR(inst, sleep_gpios, {0}),                     \
+		.en_pin = GPIO_DT_SPEC_INST_GET_OR(inst, en_gpios, {0}),                           \
+		.m0_pin = GPIO_DT_SPEC_INST_GET_OR(inst, m0_gpios, {0}),                           \
+		.m1_pin = GPIO_DT_SPEC_INST_GET_OR(inst, m1_gpios, {0}),                           \
```
These are required according to the init function :)

bjarki-andreasen reviewed (changes requested):
Could you add the driver to the stepper API test suite as well? https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/stepper/stepper_api

jbehrensnx commented:
> Could you add the driver to the stepper API test suite as well? https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/stepper/stepper_api

Done, but I had to change the overlay file from the ones used by the drv8424 tests and make a change to the test setup function to ensure compatibility. This runs into some of the same issues that will come up when making the drv8424 tests compatible with other stepper drivers.

jilaypandya commented:
Add the driver to the build_all tests.
Can you split the changes to already existing stepper_api tests into a separate commit?

jilaypandya commented:
> > Could you add the driver to the stepper API test suite as well? https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/stepper/stepper_api
> 
> Done, but I had to change the overlay file from the ones used by the drv8424 tests and make a change to the test setup function to ensure compatibility. This runs into some of the same issues that will come up when making the drv8424 tests compatible with other stepper drivers.

From what I've seen is that when the motor is not enabled then the "move" operations shall return -ECANCELLED, this is missing imo in other drivers. From what i understand, a RFC is also going to come soon i guess, with further improvements in stepper api. Looking forward to it  :-)

jbehrensnx committed:
drivers: stepper: test: Added drv8424 driver tests

Adds 2 test suites for the drv8424 stepper driver, testing API conformity
and return values in the first suite and gpio pin values in the second
suite

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

jbehrensnx commented:
> Add the driver to the build_all tests.
> Can you split the changes to already existing stepper_api into a separate commit?

Added drv8424 to build_all and put it together with the changes to stepper_api into a separate commit.

fabiobaltieri commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+
+static void drv8424_positioning_top_interrupt(const struct device *dev, void *user_data)
+{
+	struct drv8424_data *data = (struct drv8424_data *)user_data;
+	const struct drv8424_config *config = (struct drv8424_config *)data->dev->config;
```
no need for these castings, both are `void *` it's not like the compiler can do anything about it anyway, drop

fabiobaltieri commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+}
+
+static int drv8424_set_microstep_pin(const struct device *dev, const struct gpio_dt_spec *pin,
+				     int value, const char *pin_name)
```
I wouldn't bother with `pin_name`, if those gpio calls fail the whole system is messed up real bad and knowing which pin failed does not really give you much, I would drop this, save few bytes of flash and ram

fabiobaltieri commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	int m0_value = data->pin_states.m0;
+	int m1_value = data->pin_states.m1;
```
`uint8_t`? save few bytes on the stack

fabiobaltieri commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+		return -EINVAL;
+	}
+
+	data->counter_top_cfg.ticks = counter_us_to_ticks(config->counter, USEC_PER_SEC / freq_hz);
```
consider using `DIV_ROUND_CLOSEST` if you want the closest approximation rather than a floor

fabiobaltieri commented on drivers/stepper/ti/drv8424.c:
```diff
@@ -0,0 +1,680 @@
+	struct drv8424_data *data = dev->data;
+	int ret;
+
+	int m0_value = 0;
+	int m1_value = 0;
```
uint8_t

fabiobaltieri commented on dts/bindings/stepper/ti/ti,drv8424.yaml:
```diff
@@ -0,0 +1,64 @@
+    status = "okay";
+    compatible = "ti,drv8424";
+
+    dir-gpios = < &arduino_header 18 0 >;
+    step-gpios = < &arduino_header 19 0 >;
+    sleep-gpios = < &arduino_header 15 GPIO_ACTIVE_LOW >;
+    en-gpios  = < &arduino_header 14 0 >;
+    m0-gpios = < &mikroe_stepper_gpios 0 0 >;
+    m1-gpios = < &mikroe_stepper_gpios 1 0 >;
+    counter = < &counter2 >;
```
drop the spaces around the phandle

fabiobaltieri commented on tests/drivers/stepper/drv8424/api/boards/native_sim.overlay:
```diff
@@ -0,0 +1,49 @@
+
+#include <zephyr/dt-bindings/gpio/gpio.h>
+
+
```
nit: drop a blank line, leave just one

fabiobaltieri commented on tests/drivers/stepper/drv8424/api/boards/native_sim.overlay:
```diff
@@ -0,0 +1,49 @@
+		status = "okay";
+		compatible = "ti,drv8424";
+
+
```
same

fabiobaltieri commented on tests/drivers/stepper/drv8424/api/boards/native_sim.overlay:
```diff
@@ -0,0 +1,49 @@
+
+	};
+
+
```
same, you can use them between the gpio nodes if you really want to, there's normally one there :-)

fabiobaltieri commented on tests/drivers/stepper/drv8424/api/boards/nucleo_f767zi.overlay:
```diff
@@ -0,0 +1,67 @@
+};
+
+/ {
+
```
drop

fabiobaltieri commented on tests/drivers/stepper/drv8424/api/boards/nucleo_f767zi.overlay:
```diff
@@ -0,0 +1,67 @@
+		 status = "okay";
+		 compatible = "ti,drv8424";
+
+
```
drop

fabiobaltieri commented on tests/drivers/build_all/stepper/gpio.dtsi:
```diff
@@ -23,3 +23,16 @@ test_tmc2209: tmc2209_motor {
+	en-gpios  = <&test_gpio 0 0>;
+	m0-gpios = <&test_gpio 0 0>;
+	m1-gpios = <&test_gpio 0 0>;
+	counter = < &counter0 >;
```
drop the spaces around the phandle

fabiobaltieri commented on tests/drivers/stepper/stepper_api/boards/nucleo_f767zi.overlay:
```diff
@@ -0,0 +1,41 @@
+		 status = "okay";
+		 compatible = "ti,drv8424";
+
+
```
same as with the other files, just keep them cherent

fabiobaltieri commented on tests/drivers/stepper/stepper_api/boards/nucleo_f767zi.overlay:
```diff
@@ -0,0 +1,41 @@
+		 #address-cells = <1>;
+		 #size-cells = <0>;
+		 #stepper-motor-cells = <0>;
+
```
drop

fabiobaltieri reviewed (commented):
heya, few more comments, mainly styling nitpicks

jbehrensnx commented:
Sorry, made error in this push

jbehrensnx committed:
drivers: stepper: test: Added drv8424 to stepper_api and build_all tests

Adds the drv8424 stepper driver to the build_all/stepper test and the
stepper_api test. The later test is also modified to ensure compatibility.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

fouge commented:
Hey @jbehrensnx @Irockasingranite

do you know if this driver is compatible with these:
– DRV8426: 33-V, 900 mΩ HS + LS
– DRV8436: 48-V, 900 mΩ HS + LS
– DRV8434: 48-V, 330 mΩ HS + LS

should we rename the driver`drv84xx` if so? 🤔

jbehrensnx commented:
> Hey @jbehrensnx @Irockasingranite
> 
> do you know if this driver is compatible with these: – DRV8426: 33-V, 900 mΩ HS + LS – DRV8436: 48-V, 900 mΩ HS + LS – DRV8434: 48-V, 330 mΩ HS + LS
> 
> should we rename the driver`drv84xx` if so? 🤔
> 
> EDIT: I just had a look at the series, it's not that simple 😬

According to the datasheet the 26, 34 and 36 should be compatible, but we didn't have any we could test this on, so we didn't gave any statement on their compatibility.
But the rest of the DRV84xx series is complicated, some might actually work with this driver, or they might not. This would require testing on a case by case basis.
Keeping a compatibility list in _ti,drv8424.yaml_ might be an option. 
If compatibility with some of the other entries is confirmed, a rename might actually make sense, but in this a case the compatibility list would definitely be necessary.
